File: android/app/src/main/java/com/airbridge/app/transfer/interfaces/IFileTransferService.kt
package com.airbridge.app.transfer.interfaces

import com.airbridge.app.core.interfaces.ITransferSession
import com.airbridge.app.core.models.DeviceInfo

/**
 * High-level file transfer service.
 * Implemented in Iteration 4.
 */
interface IFileTransferService {
    suspend fun sendFile(filePath: String, destination: DeviceInfo): ITransferSession
    suspend fun receiveFile(sessionId: String, destinationDirectory: String): ITransferSession
    fun getActiveSessions(): List<ITransferSession>

    /**
     * Registers an inbound file-transfer message handler on [DeviceConnectionService] for the
     * given [deviceId]. Must be called once a session is established so received files are saved.
     */
    fun registerReceiveHandler(deviceId: String)
}

File: windows/AirBridge.Transfer/Interfaces/IFileTransferService.cs
using AirBridge.Core.Interfaces;
using AirBridge.Core.Models;
using AirBridge.Transport.Interfaces;
using AirBridge.Transport.Protocol;

namespace AirBridge.Transfer.Interfaces;

/// <summary>
/// High-level file transfer service. Creates and manages transfer sessions.
/// Implemented in Iteration 4.
/// </summary>
public interface IFileTransferService
{
    /// <summary>Initiates sending a file to a paired remote device.</summary>
    Task<ITransferSession> SendFileAsync(string filePath, DeviceInfo destination, CancellationToken cancellationToken = default);

    /// <summary>
    /// Accepts an incoming file transfer from a remote device.
    /// <paramref name="destinationDirectory"/> is where the file will be saved.
    /// </summary>
    Task<ITransferSession> ReceiveFileAsync(string sessionId, string destinationDirectory, CancellationToken cancellationToken = default);

    /// <summary>Returns all active or recent transfer sessions.</summary>
    IReadOnlyList<ITransferSession> GetActiveSessions();

    /// <summary>
    /// Returns a message handler that processes inbound FILE_TRANSFER_START/CHUNK/END messages
    /// and saves received files to the AirBridge Downloads folder.
    /// Register with <c>DeviceConnectionService.AddMessageHandler</c> after a session is established.
    /// </summary>
    Func<ProtocolMessage, Task> CreateReceiveHandler();

    /// <summary>
    /// Sets the active outbound channel for <see cref="SendFileAsync"/>.
    /// Call with <see langword="null"/> when the device disconnects.
    /// </summary>
    void SetChannel(IMessageChannel? channel);

    /// <summary>
    /// Clears the active channel only if it is still <paramref name="expected"/>.
    /// Safe to call from a disconnected session's finally block without wiping a newly registered channel.
    /// </summary>
    void ClearChannel(IMessageChannel expected);
}
